src/rust/plugin-registry/src/server/service.rs

```diff
@@ -109,16 +119,26 @@ pub struct PluginRegistry {
     config: PluginRegistryServiceConfig,
 }
 
-impl PluginRegistry {
-    fn ensure_artifact_size_limit(&self, bytes: &[u8]) -> Result<(), PluginRegistryServiceError> {
-        let limit_mb = &self.config.artifact_size_limit_mb;
-        if bytes.len() > (limit_mb * 1024 * 1024) {
-            Err(PluginRegistryServiceError::ArtifactTooLargeError(format!(
-                "Artifact exceeds {limit_mb}MB"
-            )))
-        } else {
-            Ok(())
+fn io_input_error(input: impl std::fmt::Display) -> std::io::Error {
+    std::io::Error::new(std::io::ErrorKind::InvalidInput, input.to_string())
+}
+
+/// Keep adding the size of an incoming chunk to an AtomicUsize.
+/// Start throwing errors when it exceeds limit.
+async fn accumulate_stream_size_to_limit(
```

Comment:
This function doesn't look like it calls `.await` anywhere internally, so I don't think it needs to be async.

Reply:
It does! In fact, I'm pretty sure the rust linter doesn't allow you to declare an un-consumed future.

However, the mechanism is unconventional.

```
        let body_stream = body_stream.then(move |result| {
            accumulate_stream_size_to_limit(stream_length_handle.clone(), limit_bytes, result)
        });
```

this replaces a `.then(move |result| async move { })` 
and `then` takes an F: [FnMut](https://doc.rust-lang.org/nightly/core/ops/function/trait.FnMut.html)(<St as [Stream](https://docs.rs/futures/latest/futures/stream/trait.Stream.html)>::[Item](https://docs.rs/futures/latest/futures/stream/trait.Stream.html#associatedtype.Item)) -> Fut